Incorrect type inference leading to false positive

### Bug description

```python
import pandas as pd


def function1():
    """
    some function declared anywhere else in the codebase.
    in another file and never called in my case.
    """
    df = pd.DataFrame()
    df.columns = [1, 2, 3]


def function2():
    """
    some function which sometimes operates on
    a default constructed DataFrame object
    """
    df = pd.DataFrame()
    # pylint incorrectly infers the type of `k` to be int
    df.rename(columns={k: k.lower() for k in df.columns})
```


### Configuration

_No response_

### Command used

```shell
pylint repro.py
```


### Pylint output

```python
************* Module repro
repro.py:20:26: E1101: Instance of 'int' has no 'lower' member (no-member)
```


### Expected behavior

No diagnostic

### Pylint version

```shell
pylint 3.0.3
astroid 3.0.2
Python 3.12.1 | packaged by conda-forge | (main, Dec 23 2023, 08:03:24) [GCC 12.3.0]
```


### OS / Environment

_No response_

### Additional dependencies

```python
pandas 2.1.4 or 1.5.3
```

Thanks for the report. I'm hearing two requests. One is to distinguish:

```py
def function2():
    df = SomeType()
    return {k: k.lower() for k in df.columns}
```

from this, which is an error:
```py
def function2():
    df = function1(df)  # assume function1 returns `df`
    return {k: k.lower() for k in df.columns}
```

That's not really on pylint's roadmap, which I know might be frustrating, but pylint/astroid's inference capabilities advertise 'all the values your variables might take' -- in other words, we've inherited a system which collects all the possible values for `SomeType.columns` and doesn't know what functions you've called on them by the time we lint one of the instances you have in hand.

The other request I'm hearing is to make this more deterministic. That's totally reasonable, but I'm having trouble reproducing.

```
% cat a.py
class SomeType:
    def __init__(self):
        self.columns: list[str] = []

def function2():
    df = SomeType()
    return {k: k.lower() for k in df.columns}

% cat b.py
from .a import SomeType

def function1():
    de = SomeType()
    de.columns = [1, 2, 3]

% pylint a.py b.py
************* Module pylint.a
a.py:7:15: E1101: Instance of 'int' has no 'lower' member (no-member)

------------------------------------------------------------------
Your code has been rated at 5.00/10 (previous run: 5.00/10, +0.00)
````

```
$ pylint a.py
$ pylint b.py
$ pylint a.py b.py
************* Module a
a.py:7:15: E1101: Instance of 'int' has no 'lower' member (no-member)
```

The issue is that when linted in isolation these files produce no diagnostic, but when linted together they do.

```
pylint 3.0.3
astroid 3.0.3
Python 3.11.6 | packaged by conda-forge | (main, Oct  3 2023, 10:40:35) [GCC 12.3.0]
```

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -77,6 +77,9 @@ repos:
         entry: pylint
         language: system
         types: [python]
+        # Not that problematic to run in parallel see Pre-commit
+        # integration in the doc for details
+        # require_serial: true
         args: ["-rn", "-sn", "--rcfile=pylintrc", "--fail-on=I"]
         exclude: tests(/\w*)*/functional/|tests/input|tests(/\w*)*data/|doc/
       # We define an additional manual step to allow running pylint with a spelling

--- a/doc/development_guide/contributor_guide/release.rst
+++ b/doc/development_guide/contributor_guide/release.rst
@@ -55,7 +55,7 @@ For example:
    scratch.
 -  Delete the ``maintenance/X.Y-1.x`` branch. (For example:
    ``maintenance/2.3.x``)
--  Select all the issues labelled ``backport maintenance/X.Y-1.x`` and
+-  Select all the *closed* issues labelled ``backport maintenance/X.Y-1.x`` and
    label them ``backported``, then rename the
    ``backport maintenance/X.Y-1.x`` label to
    ``backport maintenance/X.Y.x`` (for example rename

--- a/doc/faq.rst
+++ b/doc/faq.rst
@@ -67,6 +67,28 @@ to not be included as default messages.
 You can see the plugin you need to explicitly :ref:`load in the technical reference
 <user_guide/checkers/extensions:optional checkers>`.
 
+I want to use pylint on each keystroke in my IDE, how can I do that ?
+---------------------------------------------------------------------
+
+Don't do it: pylint's full suite of checks is not fast enough for that and never
+will be. pylint is best suited for linting on save for small projects, or for a continuous
+integration job or a git ``pre-push`` hook for big projects. The larger your repository
+is, the slower pylint will be.
+
+If you want to make pylint faster for this type of use case, you can use the ``--errors-only``
+option, which will remove all the refactor, convention, and warning checks. You can also disable
+checks with inherently high complexity that need to analyse the full code base like
+``duplicate-code`` or ``cyclic-import`` (this list is not exhaustive).
+
+Why do I have non-deterministic results when I try to parallelize pylint ?
+--------------------------------------------------------------------------
+
+pylint should analyse all your code at once in order to best infer the
+actual values that result from calls. If only some of the files are given, pylint might
+miss a particular value's type and produce inferior inference for the subset. Parallelization
+of pylint is not easy; we also discourage the use of the ``-j`` option if this matters to you.
+
+
 Which messages should I disable to avoid duplicates if I use other popular linters ?
 ------------------------------------------------------------------------------------
 

--- a/doc/user_guide/installation/pre-commit-integration.rst
+++ b/doc/user_guide/installation/pre-commit-integration.rst
@@ -3,10 +3,28 @@
 Pre-commit integration
 ======================
 
-``pylint`` can be used as a `pre-commit <https://pre-commit.com>`_ hook.
+``pylint`` can be used as a `pre-commit <https://pre-commit.com>`_ hook. We however
+discourage it as pylint -- due to its speed -- is more suited to a continuous integration
+job or a git ``pre-push`` hook, especially if your repository is large.
 
 Since ``pylint`` needs to import modules and dependencies to work correctly, the
-hook only works with a local installation of ``pylint`` (in your environment).
+hook only works with a local installation of ``pylint`` (in your environment). It means
+it can't be used with ``pre-commit.ci``, and you will need to add the following to your
+``.pre-commit-config.yaml`` ::
+
+.. sourcecode:: yaml
+
+    ci:
+      skip: [pylint]
+
+Another limitation is that pylint should analyse all your code at once in order to best infer the
+actual values that result from calls. If only some of the files are given, pylint might
+miss a particular value's type and produce inferior inference for the subset. Since pre-commit slices
+the files given to it in order to parallelize the processing, the result can be degraded.
+It can also be unexpectedly different when the file set changes because the new slicing can change
+the inference. Thus the ``require_serial`` option should be set to ``true`` if correctness and determinism
+are more important than parallelization to you.
+
 If you installed ``pylint`` locally it can be added to ``.pre-commit-config.yaml``
 as follows:
 
@@ -19,6 +37,7 @@ as follows:
         entry: pylint
         language: system
         types: [python]
+        require_serial: true
         args:
           [
             "-rn", # Only display messages

--- a/doc/user_guide/usage/run.rst
+++ b/doc/user_guide/usage/run.rst
@@ -153,6 +153,7 @@ Parallel execution
 It is possible to speed up the execution of Pylint. If the running computer
 has more CPUs than one, then the work for checking all files could be spread across all
 cores via Pylints's sub-processes.
+
 This functionality is exposed via the ``-j`` command-line parameter.
 If the provided number is 0, then the total number of CPUs will be autodetected and used.
 
@@ -163,6 +164,15 @@ Example::
 This will spawn 4 parallel Pylint sub-process, where each provided module will
 be checked in parallel. Discovered problems by checkers are not displayed
 immediately. They are shown just after checking a module is complete.
+
+You can also do your own parallelization by launching pylint multiple times on subsets
+of your files (like ``pre-commit`` with the default ``require_serial=false`` does).
+Be aware, though: pylint should analyse all your code at once in order to best infer
+the actual values that result from calls. If only some of the files are given, pylint
+might miss a particular value's type and produce inferior inference for the subset.
+It can also be unexpectedly different when the file set changes because the new
+slicing can change the inference. So, don't do this if correctness and determinism
+are important to you.
 
 Exit codes
 ----------

